nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -529,4 +542,525 @@ impl DataStore {
+    /// currently held by the provided `saga_lock_id`. If the lock is held by a
+    /// different saga UUID, the instance will remain locked. If the instance
+    /// has already been unlocked, this method will return `false`.
+    ///
+    /// # Arguments
+    ///
+    /// - `authz_instance`: the instance to attempt to unlock
+    /// - `current_gen`: the generation number for the lock being released.
+    /// - `saga_lock_id`: the UUID of the saga that's releasing the lock on this
+    ///   instance
+    pub async fn instance_updater_unlock(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        current_gen: Generation,
+        saga_lock_id: Uuid,
+    ) -> Result<bool, Error> {
+        use db::schema::instance::dsl;
+
+        let instance_id = authz_instance.id();
+
+        let result = diesel::update(dsl::instance)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(instance_id))
+            // Only unlock the instance if:
+            // - the provided updater ID matches that of the saga that has
+            //   currently locked this instance.
+            .filter(dsl::updater_id.eq(Some(saga_lock_id)))
+            // - the provided updater generation matches the current updater
+            //   generation.
+            .filter(dsl::updater_gen.eq(current_gen))
+            .set((
+                dsl::updater_gen.eq(Generation(current_gen.0.next())),
+                dsl::updater_id.eq(None::<Uuid>),
+            ))
+            .check_if_exists::<Instance>(instance_id)
+            .execute_and_check(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
+                        ResourceType::Instance,
+                        LookupType::ById(instance_id),
+                    ),
+                )
+            })?;
+
+        match result {
+            // If we updated the record, the lock has been released! Return
+            // `Ok(true)` to indicate that we released the lock successfully.
+            UpdateAndQueryResult { status: UpdateStatus::Updated, .. } => {
+                Ok(true)
+            }
+            // The generation has advanced past the generation at which the
+            // lock was held. This means that we have already released the
+            // lock. Return `Ok(false)` here for idempotency.
+            UpdateAndQueryResult {
+                status: UpdateStatus::NotUpdatedButExists,
+                ref found,
+            } if found.runtime_state.updater_gen > current_gen => Ok(false),
+            // The instance exists, but the lock ID doesn't match our lock ID.
+            // This means we were trying to release a lock we never held, whcih
+            // is almost certainly a programmer error.
```

Comment:
Or that someone stole the lock from the logically correct owner. (Should we also assert/log an error in `instance_updater_lock` if it observes that its attempt to lock returned `UpdateStatus::Updated` but the lock owner wasn't previously NULL?)

Reply:
We could probably just make `.filter(dsl::updater_id.is_null())` a condition of the query as well, honestly. The generation number *should* guard it, but there's probably no reason *not* to also check that the ID is null in the query.